vault/external_tests/api/kv_helpers_test.go

```diff
@@ -295,5 +296,137 @@ func TestKVHelpers(t *testing.T) {
+			t.Fatalf("metadata fields should have been updated by PutMetadata")
+		}
+
+		// now let's try a patch
+		maxVersions = 7
+		err = client.KVv2(mountPath).PatchMetadata(context.Background(), noDataSecretPath, api.KVMetadataInput{
+			MaxVersions:    &maxVersions,
+			CustomMetadata: map[string]interface{}{"ape": nil, "rat": "brown"},
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// check that the metadata was only partially replaced
+		md3, err := client.KVv2(mountPath).GetMetadata(context.Background(), noDataSecretPath)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if md2.CASRequired != md3.CASRequired || md2.DeleteVersionAfter != md3.DeleteVersionAfter {
+			t.Fatalf("expected fields to remain unchanged but they were updated")
+		}
+		if md3.MaxVersions == 0 {
+			t.Fatalf("field was reset to its zero value when it should not have been")
+		}
+		if md2.MaxVersions == md3.MaxVersions {
+			t.Fatalf("expected field to be updated but it remained unchanged")
+		}
+
+		// let's check the custom metadata was updated correctly
+		if r, ok := md3.CustomMetadata["rat"]; ok {
+			if r != "brown" {
+				t.Fatalf("expected value to be \"brown\"")
+			}
+		} else {
+			t.Fatalf("expected there to be a new \"rat\" key")
+		}
+
+		if _, ok := md3.CustomMetadata["ape"]; ok {
+			t.Fatalf("expected \"ape\" key to be removed")
+		}
+
+		if _, ok := md3.CustomMetadata["cat"]; !ok {
+			t.Fatalf("did not expect \"cat\" key to be removed")
+		}
+
+		// now let's do another patch to test the "explicit zero value" use case
+		explicitFalse := false
+		explicitZero := 0
+		explicitTimeZero, err := time.ParseDuration("0s")
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = client.KVv2(mountPath).PatchMetadata(context.Background(), noDataSecretPath, api.KVMetadataInput{
+			CASRequired:        &explicitFalse,
+			MaxVersions:        &explicitZero,
+			DeleteVersionAfter: &explicitTimeZero,
+			CustomMetadata:     map[string]interface{}{},
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// check that those fields were reset to their zero value
+		md4, err := client.KVv2(mountPath).GetMetadata(context.Background(), noDataSecretPath)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if len(md4.CustomMetadata) > 0 {
+			t.Fatalf("expected empty map to cause deletion of all custom metadata")
+		}
+
+		if md4.MaxVersions != 0 || md4.CASRequired != false {
+			t.Fatalf("expected fields to be reset to their zero values but they were %d and %t instead", md4.MaxVersions, md4.CASRequired)
+		}
+
+		// TODO: Bring this test back to life once the bug is fixed where we can't actually reset delete-version-after to zero...
+		// if md4.DeleteVersionAfter.String() != "0s" {
+		// 	t.Fatalf("expected delete-version-after to be reset to its zero value but instead it was %d", md4.DeleteVersionAfter.String())
+		// }
+
+		////
```

Comment:
Totally agree, it's getting unwieldy. I will refactor it.

Reply:
Okay, so I went off to change this into multiple tests with a `TestMain` function to initialize the TestVaultCluster (since spinning that up and taking that down is what takes the longest, so I want to make sure each test shares the same test cluster). But then I realized `vault.NewTestCluster` and `vault.TestWaitActive` depends on `t`, which TestMain doesn't have access to. 

So what I decided to do is just split this single massive "kv v2 helpers" test up into more `t.Run`s. That way it's hopefully a little easier to read and they can still rely on the same shared test cluster setup.